src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,774 @@
+                // (see https://docs.microsoft.com/dotnet/csharp/language-reference/keywords/try-finally, third
+                // paragraph). If we fail to renew the leases, multiple workers could be processing the same change
+                // data, but we have functionality in place to deal with this (see design doc).
+                this._logger.LogError($"Failed to renew leases due to error: {e.Message}");
+            }
+            finally
+            {
+                if (this._state == State.ProcessingChanges)
+                {
+                    // Do we want to update this count even in the case of a failure to renew the leases? Probably,
+                    // because the count is simply meant to indicate how much time the other thread has spent processing
+                    // changes essentially.
+                    this._leaseRenewalCount += 1;
+
+                    // If this thread has been cancelled, then the _cancellationTokenSourceExecutor could have already
+                    // been disposed so shouldn't cancel it.
+                    if (this._leaseRenewalCount == MaxLeaseRenewalCount && !token.IsCancellationRequested)
+                    {
+                        this._logger.LogWarning("Call to execute the function (TryExecuteAsync) seems to be stuck, so it is being cancelled");
+
+                        // If we keep renewing the leases, the thread responsible for processing the changes is stuck.
+                        // If it's stuck, it has to be stuck in the function execution call (I think), so we should
+                        // cancel the call.
+                        this._cancellationTokenSourceExecutor.Cancel();
+                        this._cancellationTokenSourceExecutor = new CancellationTokenSource();
+                    }
+                }
+
+                // Want to always release the lock at the end, even if renewing the leases failed.
+                this._rowsLock.Release();
+            }
+        }
+
+        /// <summary>
+        /// Resets the in-memory state of the change monitor and sets it to start polling for changes again.
+        /// </summary>
+        /// <param name="error">
+        /// The error messages the logger will report describing the reason function execution failed (used only in the case of a failure).
+        /// </param>
+        /// <param name="acquireLock">True if ClearRowsAsync should acquire the "_rowsLock" (only true in the case of a failure)</param>
+        private async Task ClearRowsAsync(string error, bool acquireLock)
+        {
+            if (acquireLock)
+            {
+                this._logger.LogError(error);
+                await this._rowsLock.WaitAsync();
+            }
+
+            this._leaseRenewalCount = 0;
+            this._state = State.CheckingForChanges;
+            this._rows = new List<IReadOnlyDictionary<string, string>>();
+            this._rowsLock.Release();
```

Comment:
Shouldn't this only be done if `acquireLock` is true?

Reply:
None of these are my changes so I can only lookup and guess. Here, the code tries to maintain that whenever `acquireLock` is passed as `false`, the lock is already acquired so that it is not tried to be re-acquired which would result in a deadlock (?). In this case, the first line inside method `ReleaseLeasesAsync` has acquired the lock, so it's safe to update the shared variables.

Reply:
Yeah, it'll deadlock if called a second time. 

Make an issue to look into this please - at the very least I'd like to try and refactor this slightly to make this easier to understand. All this locking stuff I want to go over anyways since right now it seems kind of arbitrary where the locks are acquired/released which could easily lead to issues later on.

Reply:
Created #286 for this. It is not clear to me right now how to not share `_this.rows` between get-changes and process-changes loops. This might require some redesigning (maybe only have the second loop manage the leases and versions and the first loop get and process rows?).